src/main/java/org/opensearch/ad/ratelimit/CheckpointReadWorker.java

```diff
@@ -213,6 +229,22 @@ protected MultiGetRequest toBatchRequest(List<EntityFeatureRequest> toProcess) {
                 }
             }
 
+            // deal with failures that we will retry for a limited amount of times
+            // before stopping the detector
+            if (stopDetectorRequests != null) {
+                for (EntityRequest origRequest : toProcess) {
+                    Optional<String> modelId = origRequest.getModelId();
+                    if (modelId.isPresent() && stopDetectorRequests.containsKey(modelId.get())) {
```

Comment:
Why cant we just loop over stopDetectorRequests instead of toProcess?

Reply:
because we need detector id from toProcess' elements.   stopDetectorRequests only has model id.